File: labs/crdt/pkg/crdt/gcounter.go
package crdt

import "maps"

// GCounter is a grow-only counter CRDT.
//
// Each node in the distributed system has its own entry in the counter map.
// A node can only increment its own entry. The global value is the sum of all
// entries. Merging two GCounters takes the max per node — this ensures
// convergence regardless of merge order.
//
// Invariants:
//   - Values never decrease (grow-only)
//   - Merge is commutative, associative, and idempotent
type GCounter struct {
	// counts maps nodeID -> count for that node.
	// A node may only increment its own entry.
	counts map[string]int64
}

// NewGCounter creates an empty GCounter.
func NewGCounter() GCounter {
	return GCounter{counts: make(map[string]int64)}
}

// Increment increments the counter for the given node by 1.
func (g *GCounter) Increment(nodeID string) {
	g.counts[nodeID]++
}

// IncrementBy increments the counter for the given node by delta.
// delta must be positive; negative deltas are ignored (use PNCounter).
func (g *GCounter) IncrementBy(nodeID string, delta int64) {
	if delta > 0 {
		g.counts[nodeID] += delta
	}
}

// Value returns the current sum of all node counters.
func (g *GCounter) Value() int64 {
	var total int64
	for _, v := range g.counts {
		total += v
	}
	return total
}

// NodeValue returns the counter value for a specific node.
func (g *GCounter) NodeValue(nodeID string) int64 {
	return g.counts[nodeID]
}

// Merge merges another GCounter into this one.
// For each node, the maximum value wins. This operation is
// commutative, associative, and idempotent.
func (g *GCounter) Merge(other GCounter) {
	if g.counts == nil {
		g.counts = make(map[string]int64)
	}
	for nodeID, count := range other.counts {
		if count > g.counts[nodeID] {
			g.counts[nodeID] = count
		}
	}
}

// Clone returns a deep copy of this GCounter.
func (g GCounter) Clone() GCounter {
	return GCounter{counts: maps.Clone(g.counts)}
}

// Entries returns a copy of the internal map for inspection/serialization.
func (g GCounter) Entries() map[string]int64 {
	return maps.Clone(g.counts)
}

File: labs/crdt/java-integration/src/main/java/dev/pushkar/crdt/GCounter.java
package dev.pushkar.crdt;

import java.util.HashMap;
import java.util.Map;

/**
 * Java port of the Go GCounter CRDT.
 *
 * <p>Identical semantics to the Go implementation: each node has its own entry in
 * the map; the global value is the sum; merging takes the max per node. The JVM
 * implementation is structurally identical — CRDTs are language-agnostic.
 *
 * <p>This class is intentionally kept simple (~50 lines) to show that the core
 * CRDT logic is a data-structure algorithm, not a framework concern.
 */
public final class GCounter {

    /** nodeID -> count for that node. */
    private final Map<String, Long> counts;

    public GCounter() {
        this.counts = new HashMap<>();
    }

    private GCounter(Map<String, Long> counts) {
        this.counts = new HashMap<>(counts);
    }

    /** Increments this node's counter by 1. */
    public void increment(String nodeId) {
        counts.merge(nodeId, 1L, Long::sum);
    }

    /** Increments this node's counter by delta (must be positive). */
    public void incrementBy(String nodeId, long delta) {
        if (delta > 0) {
            counts.merge(nodeId, delta, Long::sum);
        }
    }

    /** Returns the sum of all node counters. */
    public long value() {
        return counts.values().stream().mapToLong(Long::longValue).sum();
    }

    /**
     * Merges another GCounter into this one.
     * For each node, the maximum value wins.
     * Commutative, associative, and idempotent.
     */
    public void merge(GCounter other) {
        other.counts.forEach((nodeId, count) ->
            counts.merge(nodeId, count, Math::max)
        );
    }

    /** Returns a defensive copy of the internal map for inspection. */
    public Map<String, Long> entries() {
        return Map.copyOf(counts);
    }

    /** Returns a deep copy of this GCounter. */
    public GCounter copy() {
        return new GCounter(counts);
    }

    @Override
    public String toString() {
        return "GCounter{value=" + value() + ", entries=" + counts + "}";
    }
}
